### PR TITLE
feat: fs events

### PR DESCRIPTION
**Pull request by bartlomieju**
Prototype for file watching. 

Uses https://docs.rs/notify/5.0.0-pre.2/notify/index.html

I used some code from @jcao219

### CODE REVIEW

File: cli/ops/fs.rs [hunk 1]
```diff
@@ -577,3 +586,73 @@ fn op_cwd(
...
+  blocking_json(false, move || {
+    let mut table = state.lock_resource_table();
+    let resource = table
+      .get_mut::<FsWatcher>(args.rid as u32)
+      .ok_or_else(bad_resource)?;
+    // blocks
+    let result = resource.receiver.recv().map_err(ErrBox::from)?;
+    let event = result.map_err(ErrBox::from)?;
+    let serialized = serde_json::to_string(&event).expect("Foobar");
+    Ok(json!(serialized))
```
Comment (bartlomieju) [inline cli/ops/fs.rs]: Note to self - this is bad, it obviously blocks while holding a mutex. This must be rewritten into a `Future` using probably `try_recv` but then a waker is needed - to be figured

**Comment by bartlomieju** [comment]
After more digging I'm not sure fs events can solved elegantly at the moment. There's no way to turn this sync API into async one. 

Once we land Tokio 0.2 there's `spawn_blocking` function that can be coupled with `recv_timeout`. So postponing this PR until #3387 is done

**Comment by bartlomieju** [comment]
Not sure how to test it yet... [Docs state that contents](https://docs.rs/notify/5.0.0-pre.2/notify/event/enum.EventKind.html#variant.Access) of event depend on the platform.

Example output for `fs_watch_test.ts` on macOS:
```
{"type":{"create":{"kind":"folder"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7"],"attrs":{}}
{"type":{"create":{"kind":"file"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"content"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file1.txt"],"attrs":{}}
{"type":{"create":{"kind":"file"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"content"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file2.txt"],"attrs":{}}
{"type":{"create":{"kind":"file"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file1.txt"],"attrs":{}}
{"type":{"remove":{"kind":"file"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"content"}},"paths":["/private/var/folders/4k/qds_txm54r5ffc909vqvgdkw0000gn/T/715c24c7/file1.txt"],"attrs":{}}
```

Also serialization output is really questionable in my opinion...

**Comment by ry** [comment]
I think a platform dependent test would be fine. Probably a custom integration test would be the best way to test this. Create a temp dir, add file, touch it, etc - check events are issued in JS.

**Comment by bartlomieju** [comment]
Test case:
```ts
  const testDir = await Deno.makeTempDir();
  const file1 = testDir + "/file1.txt";
  const file2 = testDir + "/file2.txt";

  const watcher = Deno.watch(testDir, { recursive: true });
  // start capturing events in the background
  await Deno.writeFile(file1, new Uint8Array([0, 1, 2]));
  await Deno.writeFile(file2, new Uint8Array([0, 1, 2]));
  await Deno.remove(file1);
  await Deno.rename(file2, file1);
  await Deno.chmod(file1, 0o666);
  const f = await Deno.open(file1);
  f.close();

  setTimeout(() => {
    watcher.close();
  }, 750);
```

ubuntu:
```
{"type":{"create":{"kind":"file"}},"paths":["/tmp/392eadc2/file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"any"}},"paths":["/tmp/392eadc2/file1.txt"],"attrs":{}}
{"type":{"access":{"kind":"close","mode":"write"}},"paths":["/tmp/392eadc2/file1.txt"],"attrs":{}}
{"type":{"create":{"kind":"file"}},"paths":["/tmp/392eadc2/file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"any"}},"paths":["/tmp/392eadc2/file2.txt"],"attrs":{}}
{"type":{"access":{"kind":"close","mode":"write"}},"paths":["/tmp/392eadc2/file2.txt"],"attrs":{}}
{"type":{"remove":{"kind":"file"}},"paths":["/tmp/392eadc2/file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"rename","mode":"from"}},"paths":["/tmp/392eadc2/file2.txt"],"attrs":{"tracker":2133}}
{"type":{"modify":{"kind":"rename","mode":"to"}},"paths":["/tmp/392eadc2/file1.txt"],"attrs":{"tracker":2133}}
{"type":{"modify":{"kind":"rename","mode":"both"}},"paths":["/tmp/392eadc2/file2.txt","/tmp/392eadc2/file1.txt"],"attrs":{"tracker":2133}}
{"type":{"modify":{"kind":"metadata","mode":"any"}},"paths":["/tmp/392eadc2/file1.txt"],"attrs":{}}
```

macOS
```
{"type":{"create":{"kind":"folder"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f"],"attrs":{}}
{"type":{"create":{"kind":"file"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file1.txt"],"attrs":{}}
{"type":{"remove":{"kind":"file"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"content"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file1.txt"],"attrs":{}}
{"type":{"create":{"kind":"file"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"rename","mode":"from"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"data","mode":"content"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"rename","mode":"from"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"metadata","mode":"ownership"}},"paths":["/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/679f3d6f/file1.txt"],"attrs":{}}
```

Windows
```
{"type":{"create":{"kind":"any"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"any"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file1.txt"],"attrs":{}}
{"type":{"create":{"kind":"any"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"any"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file2.txt"],"attrs":{}}
{"type":{"remove":{"kind":"any"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"rename","mode":"from"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file2.txt"],"attrs":{}}
{"type":{"modify":{"kind":"rename","mode":"to"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file1.txt"],"attrs":{}}
{"type":{"modify":{"kind":"any"}},"paths":["C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\9d0f8c00\\file1.txt"],"attrs":{}}
```

**Comment by ry** [comment]
feel free to make the test sloppy - i don't think nailing down the exact order is important at this stage

**Comment by bartlomieju** [comment]
> feel free to make the test sloppy - i don't think nailing down the exact order is important at this stage

Sure, what about return type from watcher? What you see above is what `notify` serializes to

**Comment by bartlomieju** [comment]
This is getting stale; the only thing left to do in it is providing some unified interface between different OSes (see https://github.com/denoland/deno/pull/3452#issuecomment-577886619).

File: cli/ops/fs.rs [hunk 2]
```diff
@@ -637,3 +651,93 @@ fn op_cwd(
...
+type FsWatcherReceiver =
+  mpsc::Receiver<Result<notify::event::Event, notify::Error>>;
+
+#[allow(dead_code)]
+struct FsWatcher {
+  watcher: RecommendedWatcher,
+  receiver: Arc<AsyncMutex<FsWatcherReceiver>>,
+}
+
+impl Resource for FsWatcher {}
```
Comment (bartlomieju) [inline cli/ops/fs.rs]: Remove

File: cli/ops/fs.rs [hunk 3]
```diff
@@ -637,3 +651,93 @@ fn op_cwd(
...
+
+#[derive(Deserialize)]
+struct OpenWatcherArgs {
+  recursive: bool,
+  paths: Vec<String>,
+}
+
+pub fn op_watch(
+  state: &ThreadSafeState,
```
Comment (bartlomieju) [inline cli/ops/fs.rs]: s/ThreadSafeState/State/

File: cli/ops/fs.rs [hunk 4]
```diff
@@ -637,3 +651,93 @@ fn op_cwd(
...
+  for path in &args.paths {
+    state.check_read(&PathBuf::from(path))?;
+    watcher.watch(path, recursive_mode)?;
+  }
+
+  let watcher_resource = FsWatcher {
+    watcher,
+    receiver: Arc::new(AsyncMutex::new(rx)),
+  };
+  let mut table = state.lock_resource_table();
```
Comment (bartlomieju) [inline cli/ops/fs.rs]: no more lock

**Comment by ry** [comment]
@bartlomieju PTAL. I think this is ready to land.

File: cli/ops/fs_events.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  }
+  let args: OpenArgs = serde_json::from_value(args)?;
+  let (sender, receiver) = mpsc::channel::<Result<FsEvent, ErrBox>>(16);
+  let sender = std::sync::Mutex::new(sender);
+  let mut watcher: RecommendedWatcher =
+    Watcher::new_immediate(move |res: Result<Event, notify::Error>| {
+      let res2 = res.map(FsEvent::from).map_err(ErrBox::from);
+      let mut sender = sender.lock().unwrap();
+      futures::executor::block_on(sender.send(res2)).expect("fs events error");
+    })?;
```
Comment (bartlomieju) [inline cli/ops/fs_events.rs]: Great that you removed separate thread here

File: cli/ops/fs_events.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  let f = poll_fn(move |cx| {
+    let resource_table = &mut state.borrow_mut().resource_table;
+    let watcher = resource_table
+      .get_mut::<FsEventsResource>(rid)
+      .ok_or_else(bad_resource)?;
+    watcher
+      .receiver
+      .poll_recv(cx)
+      .map(|maybe_result| match maybe_result {
+        Some(Ok(value)) => Ok(json!({ "value": value, "done": false })),
```
Comment (bartlomieju) [inline cli/ops/fs_events.rs]: Nice!

File: cli/js/fs_events.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  [Symbol.asyncIterator](): AsyncIterableIterator<FsEvent> {
+    return this;
+  }
+}
+
+export function fsEvents(
+  paths: string | string[],
+  options = { recursive: true }
```
Comment (nayeemrmn) [inline cli/js/fs_events.ts]: This should be written as
```ts
{
  recursive = true
}: FsEventsOptions = {}
```
for full robustness.

**Comment by magichim** [comment]
Deno.watch feature is added at this time in 0.35 version?
@bartlomieju

**Comment by bartlomieju** [comment]
Yes https://github.com/denoland/deno/releases/tag/v0.35.0

**Comment by ry** [comment]
https://deno.land/typedoc/index.html#fsevents

**Pull request merged**